src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleEqualityTests.cs

```diff
@@ -0,0 +1,2464 @@
+            var nt2Type = model.GetTypeInfo(nt2);
+            Assert.Equal("nt2", nt2.ToString());
+            Assert.Equal("(System.Byte, System.Int64)?", nt2Type.Type.ToTestDisplayString());
+            Assert.Equal("(System.Int32, System.Int64)?", nt2Type.ConvertedType.ToTestDisplayString());
+
+            verifier.VerifyIL("C.Compare", @"{
+  // Code size      217 (0xd9)
+  .maxstack  3
+  .locals init ((int, long)? V_0,
+                bool V_1,
+                System.ValueTuple<int, long> V_2,
+                (int, long)? V_3,
+                System.ValueTuple<int, long> V_4,
+                (int, int)? V_5,
+                (int, long)? V_6,
+                System.ValueTuple<int, int> V_7,
+                (byte, long)? V_8,
+                System.ValueTuple<byte, long> V_9)
+  IL_0000:  nop
+  IL_0001:  ldstr      ""{0} ""
+  IL_0006:  ldarg.0
+  IL_0007:  stloc.s    V_5
+  IL_0009:  ldloca.s   V_5
+  IL_000b:  call       ""bool (int, int)?.HasValue.get""
+  IL_0010:  brtrue.s   IL_001e
+  IL_0012:  ldloca.s   V_6
+  IL_0014:  initobj    ""(int, long)?""
+  IL_001a:  ldloc.s    V_6
+  IL_001c:  br.s       IL_0040
+  IL_001e:  ldloca.s   V_5
+  IL_0020:  call       ""(int, int) (int, int)?.GetValueOrDefault()""
+  IL_0025:  stloc.s    V_7
+  IL_0027:  ldloc.s    V_7
+  IL_0029:  ldfld      ""int System.ValueTuple<int, int>.Item1""
+  IL_002e:  ldloc.s    V_7
+  IL_0030:  ldfld      ""int System.ValueTuple<int, int>.Item2""
+  IL_0035:  conv.i8
+  IL_0036:  newobj     ""System.ValueTuple<int, long>..ctor(int, long)""
+  IL_003b:  newobj     ""(int, long)?..ctor((int, long))""
+  IL_0040:  stloc.0
+  IL_0041:  ldloca.s   V_0
+  IL_0043:  call       ""bool (int, long)?.HasValue.get""
+  IL_0048:  stloc.1
+  IL_0049:  ldarg.1
+  IL_004a:  stloc.s    V_8
+  IL_004c:  ldloca.s   V_8
+  IL_004e:  call       ""bool (byte, long)?.HasValue.get""
+  IL_0053:  brtrue.s   IL_0061
+  IL_0055:  ldloca.s   V_6
+  IL_0057:  initobj    ""(int, long)?""
+  IL_005d:  ldloc.s    V_6
+  IL_005f:  br.s       IL_0082
+  IL_0061:  ldloca.s   V_8
+  IL_0063:  call       ""(byte, long) (byte, long)?.GetValueOrDefault()""
+  IL_0068:  stloc.s    V_9
+  IL_006a:  ldloc.s    V_9
+  IL_006c:  ldfld      ""byte System.ValueTuple<byte, long>.Item1""
+  IL_0071:  ldloc.s    V_9
+  IL_0073:  ldfld      ""long System.ValueTuple<byte, long>.Item2""
+  IL_0078:  newobj     ""System.ValueTuple<int, long>..ctor(int, long)""
+  IL_007d:  newobj     ""(int, long)?..ctor((int, long))""
+  IL_0082:  stloc.3
+  IL_0083:  ldloc.1
+  IL_0084:  ldloca.s   V_3
+  IL_0086:  call       ""bool (int, long)?.HasValue.get""
+  IL_008b:  beq.s      IL_0090
+  IL_008d:  ldc.i4.0
+  IL_008e:  br.s       IL_00c8
+  IL_0090:  ldloc.1
+  IL_0091:  brtrue.s   IL_0096
+  IL_0093:  ldc.i4.1
+  IL_0094:  br.s       IL_00c8
+  IL_0096:  ldloca.s   V_0
+  IL_0098:  call       ""(int, long) (int, long)?.GetValueOrDefault()""
+  IL_009d:  stloc.2
+  IL_009e:  ldloca.s   V_3
+  IL_00a
```

Comment:
Aleksey raised a question that needs LDM input and would certainly affect this PR: 
Do custom `==`, `&` and `true` operators have any role in tuple equality?

Two views on `(a, b) == (c, d)` so far:
• It is the same as `tempA == tempC && tempB == tempD`, which can involve custom `==` and `&`. The `true` operator will come into play when you consume the end result.
• It is the same as `(bool)(tempA == tempC) && (bool)(tempB == tempD)`, which can involve custom `==` and `true` (but not `&`).